pkg/nfd-master/nfd-master.go

```diff
@@ -440,10 +452,120 @@ func (m *nfdMaster) SetLabels(c context.Context, r *pb.SetLabelsRequest) (*pb.Se
 			klog.Errorf("failed to advertise labels: %v", err)
 			return &pb.SetLabelsReply{}, err
 		}
+
+		// set taints
+		var taints []corev1.Taint
+		if m.args.EnableTaints {
+			taints = crTaints
+		}
+
+		// Call setTaints even though the feature flag is disabled. This
+		// ensures that we delete NFD owned stale taints when flag got
+		// turned off.
+		err = m.setTaints(c, cli, taints, r.NodeName)
+		if err != nil {
+			return &pb.SetLabelsReply{}, err
+		}
 	}
 	return &pb.SetLabelsReply{}, nil
 }
 
+// setTaints sets note taints and annotations based on the taints passed via
+// nodeFeatureRule custom resorce. If empty list of taints is passed, currently
+// NFD owned taints and annotations are removed from the node.
+func (m *nfdMaster) setTaints(c context.Context, cli *kubernetes.Clientset, taints []corev1.Taint, nodeName string) error {
+	// Fetch the node object.
+	node, err := m.apihelper.GetNode(cli, nodeName)
+	if err != nil {
+		return err
+	}
+
+	// De-serialize the taints annotation into Taint type for comparision below.
+	annotations := node.GetAnnotations()
+	oldTaints := []corev1.Taint{}
+	if val, ok := annotations[nfdv1alpha1.NodeTaintsAnnotation]; ok {
+		sts := strings.Split(val, ",")
+		corev1Taints, _, err := taintutils.ParseTaints(sts)
+		if err != nil {
+			return err
+		}
+		for _, ct := range corev1Taints {
+			oldTaints = append(oldTaints,
+				corev1.Taint{
+					Key:    ct.Key,
+					Value:  ct.Value,
+					Effect: ct.Effect})
+		}
+	}
+
+	// Find out the diff between the old (from taints annotation) and
+	// new taints (from NodeFeatureRule) from which we get the taints
+	// to remove from the node.
+	taintsToRemove := []corev1.Taint{}
+	for _, taint := range oldTaints {
+		if !taintutils.TaintExists(taints, &taint) {
+			taintsToRemove = append(taintsToRemove, taint)
+		}
+	}
+
+	// We have taintsToRemove that we found from the annotation and here
+	// we first check if the taints still exist on the node and if yes
+	// remove it, otherwise move on.
+	for _, taintToRemove := range taintsToRemove {
+		if taintutils.TaintExists(node.Spec.Taints, &taintToRemove) {
+			newTaints, removed := taintutils.DeleteTaint(node.Spec.Taints, &taintToRemove)
+			if !removed {
+				klog.V(1).Infof("taint %q already deleted from node", taintToRemove.ToString())
+			}
+			node.Spec.Taints = newTaints
+		}
+	}
+
+	// Add new taints.
+	if len(taints) > 0 {
+		var newNode *corev1.Node
+		oldNode := node
+		oldNodeCopy := oldNode
+		for _, taint := range taints {
+			curNewNode, _, err := taintutils.AddOrUpdateTaint(oldNodeCopy, &taint)
+			if err != nil {
+				return fmt.Errorf("failed to add %q taint on node %v", taint, node.Name)
+			}
+			newNode = curNewNode
+			oldNodeCopy = curNewNode
+			continue
+		}
+		err = controller.PatchNodeTaints(c, cli, nodeName, oldNode, newNode)
+		if err != nil {
+			return fmt.Errorf("failed to patch the node %v", node.Name)
+		}
+
+	}
```

Comment:
Good catch finding `controller.PatchNodeTaints`, looks handy and more reliable than trying to update the whole node object 👍

But AFAIU, we should delegate all taints update to this function, i.e. also just deleting old taints. No need to check for `if len(taints) > 0` and also `newNode` should've been declared already eariler.

```suggestion
	for _, taint := range taints {
		newNode, _, err := taintutils.AddOrUpdateTaint(oldNodeCopy, &taint)
		if err != nil {
			return fmt.Errorf("failed to add %q taint on node %v", taint, node.Name)
		}
	}
	err = controller.PatchNodeTaints(c, cli, nodeName, oldNode, newNode)
	if err != nil {
		return fmt.Errorf("failed to patch the node %v", node.Name)
	}
```

Reply:
> But AFAIU, we should delegate all taints update to this function, i.e. also just deleting old taints. No need to check for `if len(taints) > 0` and also `newNode` should've been declared already eariler.
Suggested changes makes sense. We can even make it a bit shorter like this...
```
	if len(taints) > 0 {
		for _, taint := range taints {
			curNewNode, _, err := taintutils.AddOrUpdateTaint(newNode, &taint)
			if err != nil {
				return fmt.Errorf("failed to add %q taint on node %v", taint, node.Name)
			}
			newNode = curNewNode
			continue
		}
		err = controller.PatchNodeTaints(context.TODO(), cli, nodeName, node, newNode)
		if err != nil {
			return fmt.Errorf("failed to patch the node %v", node.Name)
		}
	}
```
Basically, `oldNodeCopy` `oldNode` variables are not needed. Regarding  `if len(taints) > 0`, I think we will hit nil pointer if we don't use it.

Reply:
> Regarding if len(taints) > 0, I think we will hit nil pointer if we don't use it.

Why would that be? We don't have any pointers. If `taints==nil` the for loop just short-circuits and we don't try to reference `&taint`.

Reply:
sorry I meant in the tests https://github.com/kubernetes-sigs/node-feature-discovery/blob/0ec98a57d2bca625ee949950a14aec64b5fa5d79/pkg/nfd-master/nfd-master-internal_test.go#L333